### ISSUE TITLE
Allow inputs from outside {workspaceRoot}

### ISSUE THREAD

**Issue by woppa684** [nrwl/nx#12847]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
I would like to be able to 
- specify inputs from outside the workspaceRoot
- have options interpolated in inputs just as in outputs

## Motivation
My frontend is in the same GIT repo as my backend, both as a different sub folder in the root. They use the same source IDL file (XML) to generate an API. This source file is located together with the server / backend. I have created a custom executor to do this code generation. Its target definition looks something like this (some internals changed). Notice that I already included what I would think the syntax should look like.

```json
"generate": {
    "executor": "@canonpp/sdk:generate",
    "inputs": ["default", "{options.idlFile}", "{options.template}/**/*", "{options.apiIncludeDirs}", "!{options,outputPath}"],
    "options": {
        "idlFile": "../backend/api.xml",
        "template": "node_modules/@canonpp/runtime-idl/templates/interface/",
        "outputPath": "libs/api/src/.generated",
        "apiIncludeDirs": ["node_modules/@canonpp/runtime-idl/apis"]
    }
},
```

I have tried specifying inputs like `"{workspaceRoot}/../backend/api.xml". Although this doesn't throw an error, it also doesn't seem to work. I don't know whether you can already easily create a fileset from an array of folders, but this is also somehow included in the above example.

## Suggested Implementation
Stop requiring that all inputs must start with either projectRoot or workspaceRoot, interpolate options in inputs as well.

**Comment by FrozenPandaz** [comment]
Is it possible to move `nx.json` up in your repo so that it encompasses your backend and IDLs?

**Comment by woppa684** [comment]
Well, that would at least be something we would not like to have, as we would be "polluting" the root, which in our case just contains a small set of files we need for our build system configuration. I can give it a try, to see what other files need to be moved up a level and how the paths used in our project configs need to change. It would also mean that commands need to be run from the git root instead of the current Nx workspace root I guess. This would not be my ideal scenario.

I don't thinks it's weird to monitor some external files as well by the way. but if that's undesirable then a least I would like to have the options used when interpolating the inputs. For outputs this is already the case and I don't see why inputs would be different.

**Comment by woppa684** [comment]
> Is it possible to move `nx.json` up in your repo so that it encompasses your backend and IDLs?

Yeah, solved it like that. It would still be great to also interpolate inputs with options (just as outputs) since now I have to write the same path twice.

**Comment by ianldgs** [comment]
I have a similar use-case: 
The open-api specification json file is generated in the back-end folder.
There's a target to generate the typescript code based on that. Inputs are configured with the json file, but changing the json file does not invalidate the cache.
Solved by not making the open-api generation target a cacheable operation :/

**Comment by grubeninspekteur** [comment]
I have the exact same use-case as @ianldgs . As a workaround, I calculate the hash as a runtime input, but the library I found (folder-hash) takes about one second. The native hasher of Nx is probably faster, however I don't know how to expose/call it.

```"inputs": [{"runtime": "npx folder-hash ../path/to/openapi/folder"}]```

**Comment by Kaemmelot** [comment]
Same problem here with openapi-generator which uses a file from the backend project. Nx is only used for the frontend project and I cannot move the nx.json.
The solution of @grubeninspekteur does not work for me. I can see the hash of the command change but the target is not executed again, which means there is never a cache miss.

**Comment by khalilou88** [comment]
I also asked the possibility to accept `outputs` outside {workspaceRoot} for a more generic solution: https://github.com/nrwl/nx/issues/20961

**Comment by FrozenPandaz** [comment]
I think we can handle this. :thinking:.

If it's relative, from the workspace root, I think that's fine.

If it's an absolute path, I think we'll need some new indicator such as `{absolutePath}:~/.config`.

Caveat though, for the generated pipelines / Nx Cloud config, we will only clone the repo in the CI environment. If the input is not there, it may cause some unintended cache misses. You'll have to add custom configuration to ensure that the files are there.

We'll need to hash out more details before implementing but what do you all think?

**Comment by ianb-pomelo** [comment]
This is still relevant

**Comment by aivarsliepa** [comment]
My team would also benefit from this.

Is anyone working on it? If not, I could try to contribute.

**Comment by tmm1** [comment]
I tried making a symlink in my workspace to point to the outside location, but that also doesn't work.

**Comment by sebas1208** [comment]
My team will also benefit from this, any plans to add it?

**Comment by fengxi-mao-mchp** [comment]
Yes, I think this would be a useful feature. My team would also benefit from it.

**Comment by AmrAbdulrahman** [comment]
I have found a solution 🎉

```
{
  "name": "my-lib",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "my-lib/src",
  "tags": ["type:lib", "scope:api"],
  "targets": {
    "generate": { // ⚠️ this depends on external openapi.json generated by the backend outside nx workspace root
      "executor": "nx:run-commands",
      "cache": true,
      "inputs": [
        { "dependentTasksOutputFiles": "**/openapi.json" }, // ⚠️ depend here on the local copy of my external dep
        "{projectRoot}/swagger-config.json", // ⚠️ depend here on other local deps
      ],
      "outputs": ["{projectRoot}/src/lib"],
      "dependsOn": ["copy-openapi-json"], // ⚠️ copy the external json first as a dependency
      "options": {
        "commands": [
          "ng-openapi-gen --config libs/lib-customer-api/swagger-config.json" // 🟢 we don't do this now unless the backend generated openapi.json changes, or else, the target "generate" is restored from cache.
        ],
        "parallel": false
      }
    },
    "copy-openapi-json": {
      "executor": "nx:run-commands",
      "cache": false,
      "outputs": ["{projectRoot}/openapi.json"],
      "options": {
        "commands": ["cp ../../external/path/to/openapi.json local/path/my-lib/openapi.json"]
      }
    }
  }
}
```

Just add now your copied external file(s) to your local lib `.gitignore`, and you're good to go.

Hope this helps!

**Comment by Kaemmelot** [comment]
Nothing has changed afaik, this is still relevant for me.
